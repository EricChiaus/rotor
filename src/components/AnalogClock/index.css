.analog-clock {
    border-radius: 50%;
    height: 20em;
    position: relative;
    width: 20em;
    box-shadow: -16px 16px 16px 1px rgba(0, 0, 0, 0.4);
    background-color: black;
}

.hours,
.minutes,
.seconds {
    background-color: white;
    position: absolute;
}

.hours {
    height: 20%;
    left: 48.75%;
    top: 30%;
    transform-origin: 50% 100%;
    width: 2%;
}

.minutes {
    height: 40%;
    left: 49%;
    top: 10%;
    transform-origin: 50% 100%;
    width: 1%;
}

.seconds {
    height: 45%;
    left: 49.5%;
    top: 14%;
    transform-origin: 50% 80%;
    width: 0.5%;
    z-index: 2;
}
