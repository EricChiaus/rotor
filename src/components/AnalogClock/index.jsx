import React, { useEffect, useRef } from 'react';
import {
    getHoursAngle,
    getMinutesAngle,
    getSecondsAngle,
} from '../../helpers/angles';
import './index.css';

// hours: 0~23
// minutes: 0~59
// seconds: 0-59
const AnalogClock = ({ hours, minutes, seconds }) => {
    const hoursRef = useRef(null);
    const minutesRef = useRef(null);
    const secondsRef = useRef(null);

    useEffect(() => {
        hoursRef.current.style.transform = `rotateZ(${getHoursAngle(
            hours,
            minutes,
        )}deg`;
        minutesRef.current.style.transform = `rotateZ(${getMinutesAngle(
            minutes,
        )}deg`;
        secondsRef.current.style.transform = `rotateZ(${getSecondsAngle(
            seconds,
        )}deg`;
    }, [hours, minutes, seconds]);

    return (
        <div className="analog-clock">
            <div className="hours" ref={hoursRef}></div>
            <div className="minutes" ref={minutesRef}></div>
            <div className="seconds" ref={secondsRef}></div>
        </div>
    );
};

export default AnalogClock;
