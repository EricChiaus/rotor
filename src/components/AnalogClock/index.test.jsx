import { render } from '@testing-library/react';
import AnalogClock from '.';

describe('<AnalogClock />', () => {
    it('matches the snapshot', () => {
        const { container } = render(
            <AnalogClock hours={1} minutes={0} seconds={59} />,
        );
        expect(container).toMatchSnapshot();
    });
});
