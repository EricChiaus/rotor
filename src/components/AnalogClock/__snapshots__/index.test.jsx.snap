// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AnalogClock /> matches the snapshot 1`] = `
<div>
  <div
    class="analog-clock"
  >
    <div
      class="hours"
      style="transform: rotateZ(30deg;"
    />
    <div
      class="minutes"
      style="transform: rotateZ(0deg;"
    />
    <div
      class="seconds"
      style="transform: rotateZ(354deg;"
    />
  </div>
</div>
`;
